src/simsopt/field/force.py

```diff
@@ -182,92 +442,2193 @@ def mean_squared_force_pure(gamma, gammadash, gammadashdash, quadpoints, current
+
+    # Update diagonal elements
+    Lij = jnp.where(diag_mask, diag_values, Lij)
+    return 1e-7 * Lij
+
+
+def coil_coil_inductances_inv_pure(gammas, gammadashs, a_list, b_list, downsample, cross_section):
+    """
+    Pure function for computing the inverse of the coil inductance matrix.
+
+    Args:
+        gammas (array): Array of coil positions.
+        gammadashs (array): Array of coil tangent vectors.
+        a_list (array): Array of coil radii.
+        b_list (array): Array of coil thicknesses.
+        downsample (int): Downsample factor for the objective function.
+        cross_section (str): Cross-section type.
+
+    Returns:
+        array: Array of inverse of the coil inductance matrix.
+    """
+    # Lij is symmetric positive definite so has a cholesky decomposition
+    C = jnp.linalg.cholesky(coil_coil_inductances_full_pure(gammas, gammadashs, a_list, b_list, downsample, cross_section))
+    inv_C = jscp.linalg.solve_triangular(C, jnp.eye(C.shape[0]), lower=True)
+    inv_L = jscp.linalg.solve_triangular(C.T, inv_C, lower=False)
+    return inv_L
+
+
+def coil_currents_barebones(gammas, gammadashs, gammas_TF, gammadashs_TF, currents_TF, a_list, b_list, downsample, cross_section):
+    """
+    Pure function for computing the induced currents in a set of passive coils 
+    due to a set of TF coils + the other passive coils. 
+
+    Args:
+        gammas (array): Array of passive coil positions.
+        gammadashs (array): Array of passive coil tangent vectors.
+        gammas_TF (array): Array of TF coil positions.
+        gammadashs_TF (array): Array of TF coil tangent vectors.
+        currents_TF (array): Array of TF coil current.
+        a_list (array): Array of passive coil radii.
+        b_list (array): Array of passive coil thicknesses.
+        downsample (int): Downsample factor for the objective function.
+        cross_section (str): Cross-section type.
+
+    Returns:
+        array: Array of induced currents.
+    """
+    return -coil_coil_inductances_inv_pure(gammas, gammadashs, a_list, b_list, downsample, cross_section) @ net_fluxes_pure(gammas, gammadashs, gammas_TF, gammadashs_TF, currents_TF, downsample)
+
+
+def tve_pure(gamma, gammadash, gammas2, gammadashs2, current, currents2, a, b, downsample, cross_section):
+    r"""Pure function for minimizing the total vacuum energy on a coil.
+
+    The function is
+
+     .. math::
+        J = \frac{1}{2}I_iL_{ij}I_j
+
+    where :math:`L_{ij}` is the coil inductance matrix (positive definite),
+    and :math:`I_i` is the current in the ith coil.
+
+    Args:
+        gamma (array): Array of coil positions.
+        gammadash (array): Array of coil tangent vectors.
+        gammas2 (array): Array of coil positions.
+        gammadashs2 (array): Array of coil tangent vectors.
+        current (array): Array of coil current.
+        currents2 (array): Array of coil current.
+        a (float): Coil radius.
+        b (float): Coil thickness.
```

Comment:
These descriptions of a and b aren't quite accurate. For rectangular x-section coils, `a` is not a radius. `b` only matters if the x-section is rectangular right? It's unclear how "thickness" is distinct from radius.